walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,214 @@
+          if (begin.duration() >= duration){
+            splitEvents.add(begin);
+          }
+        }
+        prevEndTime = e.getWhen().end();
+
+      }
+    }
+    if (prevEndTime < range.end()) {
+      splitEvents.add(TimeRange.fromStartEnd(prevEndTime, range.end(), false));
+    }
+    return splitEvents;
+  }
+
+  /**
+   * Counts the number of optional attendees can not make it to the meeting at the time range
+   * Compares to the minimum number of optional attendees that can not make it
+   * @param bestTimes collection which has the TimeRanges of meetings that has the minOverlap attendees
+   * @param events events of the day
+   * @param optional optional attendees for the meeting
+   * @param range range in question to see how many optional attendees can not make it
+   * @param minOverlaps current minimum number of optional attendees that can not make it to the meeting at the TimeRange. 
+   * @return minimum number of overlaps. Return the minimum of current range's overlaps and minOverlaps
+   */
+  private int optionalCounter(Collection<TimeRange> bestTimes, Collection<Event> events, Collection<String> optional, TimeRange range, int minOverlaps) {
+    int counterOverlaps = 0;
+    for (String opAttendee : optional) {
+      Collection<Event> eventsWithAttendee = eventFinderWithAttendee(events, opAttendee);
+      for (Event e : eventsWithAttendee){
+        if (e.getWhen().overlaps(range)) {
+          counterOverlaps += 1;
+        }
+      }
+    }
+    if (counterOverlaps < minOverlaps) {
+      bestTimes.clear();
+      bestTimes.add(range);
+      return counterOverlaps;
+    }
+    else if (counterOverlaps == minOverlaps) {
+      bestTimes.add(range);
+    }
+    return minOverlaps;
+  }
+
+  /**
+   * Determines the TimeRanges where the mandatory and as many optional attendees as possible can attend the meeting
+   * @param events events of the day
+   * @param request includes mandatory and optional attendees, and the duration of the meeting
+   * @return collection of TimeRanges where all mandatory and the greatest number of optional attendees can be at the meeting
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+ 
+    // get mandatory and optional attendees
+    Collection<String> mandatory = request.getAttendees();
+    Collection<String> optional = request.getOptionalAttendees();
+    
+    // sort the events to be in order by start time
+    List<Event> events_list = new ArrayList<Event>(events); 
+    Collections.sort(events_list, Event.ORDER_BY_START); 
+    List<TimeRange> possibleTimes = new ArrayList<>();
+
+    // if the duration of meeting is invalid, then there are no meeting times.
+    if (duration > TimeRange.getTimeInMinutes(23, 59) || duration < 0) {
```

Comment:
Can't meetings be longer than 30 minutes? There's an option for 90 minute meetings

Reply:
the `duration > TimeRange.getTimeInMinutes(23, 59)` refers to if the duration is longer than 23 hours and 59 minutes